[WIP] [FR] add Sequential.append, .extend , .insert & .pop

codexetreme opened the pull request:
adding features as required from : #20117 
cc: @SsnL @apaszke

codexetreme committed:
starter code

codexetreme commented:
Why does it show 220 changed files, I have changed only 1 in reality

codexetreme commented:
Here, how will the functions be implemented? In Python only or in c++, such what when, say `append()` is called, it has some c++ bound function that is called

SsnL commented on torch/nn/modules/container.py:
```diff
@@ -87,6 +87,19 @@ def __dir__(self):
         return keys
 
+    # need to add code, just a starter
+    def append(self):
```
all these need to be on sequential class.

ssnl reviewed (commented):
python sequential and cpp sequential are unrelated. the cpp one seems to have a different API, so I think you should focus on the python one for now. 

cpp: https://github.com/pytorch/pytorch/blob/7bc8562a9a43af7ae779fdef22326efe8644ae92/torch/csrc/api/include/torch/nn/modules/sequential.h

codexetreme commented:
Alright, I will work on it accordingly

codexetreme committed:
added code to implement append, extend and pop

codexetreme commented on torch/nn/modules/container.py:
```diff
@@ -87,6 +87,32 @@ def __dir__(self):
+            raise ValueError('The given parameter {}, is of type {},\n expected it to be of type torch.nn.Sequential'.format(sequential,torch.typename(sequential)))
+
+    def insert(self,i,layer):
+        pass
```
For insert, since it is a ordered dict, there are some values that will be in strings that are non numeric, so how exactly can the additional layers be added? eg: say the sequential container has 2 layers `test` and `test2` and the input of `i` is 1, then where exactly  do we add it?

crcrpar replied:
I'm not sure this is OK for core developers, but the below might insert a layer at index `i`.
Note that I didn't test this snippet.

```python
    def insert(self, i, new_layer):
        new_modules = OrderedDict()
        for j, (key, layer) in enumerate(self._modules.items()):
            if i == j:
                new_modules[str(i)] = new_layer
            new_modules[key] = layer
        del self._modules
        self._modules = new_modules
```

codexetreme replied:
Ok, I will implement similarly, if the Core Devs aren't okay, I can change it :)

codexetreme committed:
bug fix

codexetreme commented:
Here, where to add the tests?

crcrpar commented on torch/nn/modules/container.py:
```diff
@@ -87,6 +87,33 @@ def __dir__(self):
         keys = [key for key in keys if not key.isdigit()]
         return keys
 
+    # need to add code, just a starter
```
Could you remove this line, please?

crcrpar commented:
> Here, where to add the tests?

At least, add tests to `test_nn.py` and my suggestion is around https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L1040.

codexetreme commented:
@crcrpar Can I just make a new test method? Since this new method will test for all possible cases like popping on an empty sequential, and so on.. (the method you linked `test_named_modules` seems to be for names modules specifically, whereas these methods are an extension) and I feel should be tested in a method of their own to check many possible cases.

crcrpar commented:
Yes, we should add new tests because methods you are implementing are new.
I wanted to suggest the location of the new tests by the link because test
files are huge.

On Tue, May 14, 2019 at 15:11 Yashodhan Ghadge <notifications@github.com>
wrote:

> @crcrpar <https://github.com/crcrpar> Can I just make a new test method?
> Since this new method will test for all possible cases like popping on an
> empty sequential, and so on.. (the method you linked test_named_modules
> seems to be for names modules specifically, whereas these methods are an
> extension) and I feel should be tested in a method of their own to check
> many possible cases.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/20274?email_source=notifications&email_token=AD3Q7U6OVOLRPUJYTDVNC53PVJJYHA5CNFSM4HLSLWO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVKLXDY#issuecomment-492092303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD3Q7U4PJP5BNBKRBGX5453PVJJYHANCNFSM4HLSLWOQ>
> .
>

codexetreme committed:
Added tests, insert functionality and fixed linting

codexetreme commented:
@crcrpar I have added the tests in a separate function as requested. Also, I added the insert functionality, but the code seems to be very hacky (since there is a need to create a new OrderedDict plus I change keys if they clash when inserting) One possibility is to maybe not have insert functionality? because from how I normally use pytorch, I would just create a new sequential and make it "insert" like so


```
def test():

    l1 = nn.Linear(2,3)
    l2 = nn.Linear(3,3)
    l3 = nn.Linear(3,3)
    l4 = nn.Linear(3,1)
    k = nn.Sequential(l1,l2,l4)
    mid_layer = nn.Sequential(l3)

    p = k[0:1](torch.tensor([2,3],dtype=torch.float))
    
    # basically I have "inserted" my layer here 
    p = mid_layer(p)
    
    p = k[-1](p)
    print (p)
```

crcrpar commented:
@SsnL @apaszke Can I get your opinion about `insert` method?

IMO, one difference between PyTorch's `Sequential` and Chainer's is that the former inherits `Module` while the latter inherits `ChainList` which maintains children layers using a Python list.

codexetreme commented:
any updates on this?

codexetreme commented:
Can anyone please give an update this? This seems to be an easy issue to fix. It'll reduce the open PR count too :)

jamaliki commented:
Hi, is there something I can contribute to on this? I was thinking of making a PR for these features but then I discovered this open one. If there is anything to add, please let me know.
